flame/trafgen.cpp

```diff
@@ -192,78 +186,30 @@ void TrafGen::start_tcp_session()
 
     // INCOMING: remote peer closed connection, EOF
     _tcp_handle->on<uvw::EndEvent>([this](uvw::EndEvent &event, uvw::TcpHandle &h) {
-        _tcp_handle->shutdown();
+        _tcp_session->on_end_event();
     });
 
     // OUTGOING: we've finished writing all our data and are shutting down
     _tcp_handle->on<uvw::ShutdownEvent>([this](uvw::ShutdownEvent &event, uvw::TcpHandle &h) {
-        _tcp_handle->close();
+        _tcp_session->on_shutdown_event();
     });
 
     // INCOMING: remote peer sends data, pass to session
     _tcp_handle->on<uvw::DataEvent>([this](uvw::DataEvent &event, uvw::TcpHandle &h) {
-        auto session = std::static_pointer_cast<TCPSession>(_tcp_handle->data());
-        session->received(event.data.get(), event.length);
+        _tcp_session->receive_data(event.data.get(), event.length);
     });
 
     // OUTGOING: write operation has finished
     _tcp_handle->on<uvw::WriteEvent>([this](uvw::WriteEvent &event, uvw::TcpHandle &h) {
-        start_wait_timer_for_tcp_finish();
+        if (!_finish_session_timer)
```

Comment:
Was this added because the assert in `start_wait_time_for_tcp_finish()` was firing? I'm wondering if it really should fail in that case (indicating an anomaly), or that should be considered nominal.

Reply:
From memory, yes, it was the assert firing. I suspect I didn't investigate why it was firing at the time; my focus was simply on getting Flamethrower working for some internal measurements.